recipes/dev/new_sft_distributed.py

```diff
@@ -0,0 +1,357 @@
+
+    
+def setup_loss(state):
+    state.loss = config.instantiate(state.loss)
+    if state.compile:
+        training.compile_loss(state.loss, verbose=state.rank==0)
+    if isinstance(state.loss, torchtune.modules.loss.CEWithChunkedOutputLoss)
+        state.model.set_num_output_chunks(state.loss.num_output_chunks)
+    utils.log_rank_zero(log, "Loss is initialized.")
+
+    # ---- cache ignore labels ---- #
+    state.ignore_labels_cache = torch.full((state.batch_size, 1), state.loss.ignore_index, device=state.device)
+
+
+def setup_data(state, dataset):
+    # [todo] add util?
+    if isinstance(dataset, ListConfig):
+        datasets = [config.instantiate(ds, state.tokenizer) for ds in dataset]
+        ds = ConcatDataset(datasets=datasets)
+        packed = getattr(ds, "packed", False)
+    else:
+        ds = config.instantiate(dataset, state.tokenizer)
+        packed = dataset.get("packed", False)
+
+    # Instantiate collate_fn
+    if "left_pad_sequence" in state.collate_fn:
+        raise RuntimeError("left_pad_sequence collator is only for inference.")
+    collate_fn = _get_component_from_path(state.collate_fn)
+
+    state.sampler = DistributedSampler(
+        ds, num_replicas=state.dp_size, rank=state.dp_rank, shuffle=state.shuffle, seed=0
+    )
+    state.dataloader = DataLoader(
+        dataset=ds,
+        batch_size=state.batch_size,
+        sampler=state.sampler,
+        # dropping last avoids shape issues with compile + flex attention
+        drop_last=True,
+        collate_fn=(
+            partial(
+                collate_fn,
+                padding_idx=state.tokenizer.pad_id,
+                ignore_idx=state.loss.ignore_index,
+            )
+            if not packed
+            else padded_collate_packed
+        ),
+    )
+
+    utils.log_rank_zero(log, "Dataset and Sampler are initialized.")
+
+
+def resume_training(state)
+    state.checkpointer.update_state(state)
+    state.checkpointer.load_distributed_checkpoint(state.model, state.optimizer)
+
+
+def eval(state):
+    # [todo] set model.eval, copy "train batch" section, copy "log section"
+
+
+def train(state):
+    training.cleanup_before_training()
+    state.optimizer.zero_grad()
+
+    with state.profiler() as profiler:
+        for state.step in range(state.step, state.total_steps):
+            # ---- train batch ---- #
+            with GradientAccumulation(state.gradient_accumulation_steps, state.model, state.data_parallel_size) as grad_acc:
+                for _ in range(grad_acc.steps):
+                    batch = next(state.dataloader)
+                    utils.batch_to_device(batch, state.device)
+
+                    # Shape [b, s], needed for the loss not the model
+                    labels = batch.pop("labels")
+                    num_tokens = (labels != loss.ignore_index).sum()
+
+                    with self.activation_offloading:
+                        logits = state.model(**batch)
```

Comment:
Tune supports packing today. Don't we need to send the mask to the model here to enable it? (Can be optional if you pad).

This is also a great example on why TensorDict/TensorClass can be ergonomical!